### ISSUE TITLE
Returning TypedReference

### ISSUE THREAD

**Issue by IllidanS4** [dotnet/roslyn#40596]
**Version Used**: 
Version 15.8.1

**Steps to Reproduce**:

```cs
public static ref T Test1<T>()
{
    return ref ((new T[1])[0]);
}

public static Span<T> Test2<T>()
{
    return new Span<T>(new T[1]);
}

public static TypedReference Test3<T>()
{
    return __makeref((new T[1])[0]);
}
```

**Expected Behavior**:
All three should compile, as they all do the same thing in essence.

**Actual Behavior**:
`Test3` fails with error CS1599: Method or delegate cannot return type 'TypedReference'.

Why values of `TypedReference` still cannot be returned? Now that there are `ref` returns and `ref` structs and there are proper checks to ensure memory safety when using these features, is there a reason these cannot be applied to a `TypedReference`? Especially considering there are methods like `TypedReference.MakeTypedReference` that seem to have no trouble returning one at all.

The same rules should apply to `RuntimeArgumentHandle` and `ArgIterator` as well.

**Comment by gafter** [comment]
Your first two methods use C# language features that are specified to work as they do.  `__makeref` is an unsafe hack and has not had language rules defined for it to make it safe.

If you want language rules added to make returning a value constructed from `__makeref` safe, please submit your feature request to [`csharplang`](https://github.com/dotnet/csharplang).

**Comment by IllidanS4** [comment]
@gafter I do not agree. Returning `TypedReference` doesn't work even if produced by other means than `__makeref`:
```cs
public static TypedReference GetTypedReference()
{
    return TypedReference.MakeTypedReference(null, null);
}
```
I tried to find any mentions of `TypedReference` or other types in the specification and without luck. Thus I ask here why it is treated differently when compared to `Span<T>` and other types. Because if `TypedReference` is a `ref struct`, the specification mandates that it should behave like all other types of this kind. The compiler makes the difference here, not the specification.

**Comment by gafter** [comment]
@IllidanS4 The compiler is doing what it was designed to do.  If you would like the language to behave differently, please submit a proposal on `csharplang`.  Since a `TypedReference` can be created in ways that expose references that are not `ref` parameters (e.g. your original example), any [specification for how to make it safe](https://github.com/dotnet/csharplang/blob/master/proposals/csharp-7.2/span-safety.md) will have to take that into account.  With the specification of `ref struct` and its safety rules as they are today, permitting the return of a `TypedReference` would permit code that exposes a reference past its lifetime.

In short, the compiler is behaving as designed.
